From ef8aa31cd460bd8fa30721b377f2068b03617129 Mon Sep 17 00:00:00 2001
From: Igor Russkikh <irusskik@redhat.com>
Date: Fri, 6 Nov 2020 18:37:20 -0500
Subject: [PATCH 063/139] [netdrv] net: atlantic: possible fault in transition
 to hibernation

Message-id: <1604687916-15087-64-git-send-email-irusskik@redhat.com>
Patchwork-id: 338489
Patchwork-instance: patchwork
O-Subject: [RHEL8.4 BZ 1857861 063/139] net: atlantic: possible fault in transition to hibernation
Bugzilla: 1857861
RH-Acked-by: David Arcari <darcari@redhat.com>
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: http://bugzilla.redhat.com/1857861

commit 52a22f4d6ff95e8bdca557765c04893eb5dd83fd
Author: Pavel Belous <pbelous@marvell.com>
Date:   Fri Feb 14 18:44:57 2020 +0300

    net: atlantic: possible fault in transition to hibernation

    during hibernation freeze, aq_nic_stop could be invoked
    on a stopped device. That may cause panic on access to
    not yet allocated vector/ring structures.

    Add a check to stop device if it is not yet stopped.

    Similiarly after freeze in hibernation thaw, aq_nic_start
    could be invoked on a not initialized net device.
    Result will be the same.

    Add a check to start device if it is initialized.
    In our case, this is the same as started.

    Fixes: 8aaa112a57c1 ("net: atlantic: refactoring pm logic")
    Signed-off-by: Pavel Belous <pbelous@marvell.com>
    Signed-off-by: Nikita Danilov <ndanilov@marvell.com>
    Signed-off-by: Igor Russkikh <irusskikh@marvell.com>
    Signed-off-by: Dmitry Bogdanov <dbogdanov@marvell.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Igor Russkikh <irusskik@redhat.com>

Cc: David Arcari <darcari@redhat.com>
Cc: Igor Russkikh <irusskik@redhat.com>
Signed-off-by: Jan Stancek <jstancek@redhat.com>
---
 drivers/net/ethernet/aquantia/atlantic/aq_pci_func.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/aquantia/atlantic/aq_pci_func.c b/drivers/net/ethernet/aquantia/atlantic/aq_pci_func.c
index 2bb329606794..f74952674084 100644
--- a/drivers/net/ethernet/aquantia/atlantic/aq_pci_func.c
+++ b/drivers/net/ethernet/aquantia/atlantic/aq_pci_func.c
@@ -359,7 +359,8 @@ static int aq_suspend_common(struct device *dev, bool deep)
 	netif_device_detach(nic->ndev);
 	netif_tx_stop_all_queues(nic->ndev);
 
-	aq_nic_stop(nic);
+	if (netif_running(nic->ndev))
+		aq_nic_stop(nic);
 
 	if (deep) {
 		aq_nic_deinit(nic, !nic->aq_hw->aq_nic_cfg->wol);
@@ -375,7 +376,7 @@ static int atl_resume_common(struct device *dev, bool deep)
 {
 	struct pci_dev *pdev = to_pci_dev(dev);
 	struct aq_nic_s *nic;
-	int ret;
+	int ret = 0;
 
 	nic = pci_get_drvdata(pdev);
 
@@ -390,9 +391,11 @@ static int atl_resume_common(struct device *dev, bool deep)
 			goto err_exit;
 	}
 
-	ret = aq_nic_start(nic);
-	if (ret)
-		goto err_exit;
+	if (netif_running(nic->ndev)) {
+		ret = aq_nic_start(nic);
+		if (ret)
+			goto err_exit;
+	}
 
 	netif_device_attach(nic->ndev);
 	netif_tx_start_all_queues(nic->ndev);
-- 
2.13.6

